documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc

```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
 ----
 ====
 
+[[passing-context]]
+=== Passing context or state objects to custom methods
+
+Additional _context_ or _state_ information can be passed through generated mapping methods to custom methods with `@Context` parameters. Such parameters are passed to other mapping methods, `@ObjectFactory` methods (see <<object-factories>>) or `@BeforeMapping` / `@AfterMapping` methods (see <<customizing-mappings-with-before-and-after>>) when applicable and can thus be used in custom code.
+
+The type of an `@Context` parameter is also inspected for factory methods and `@BeforeMapping` / `@AfterMapping` methods, which are called on the provided context parameter value if applicable.
+
+*Note:* no `null` checks are performed before calling factory or before/after mapping methods on context parameters. The caller needs to make sure that no `null` are passed in that case.
+
+For generated code to call a method that is declared with `@Context` parameters, the declaration of the mapping method being generated needs to contain at least those (or assignable) `@Context` parameters as well. MapStruct will not create new instances of missing `@Context` parameters nor will it pass `null` instead.
```

Comment:
Can you put one sentence per line? That makes it easier to do diffs down the road.

"MapStruct will" sounds as if it s a runtime component. Better change it into something like "Code generated by MapStruct will not...".